`textDocument` has an illegal filesystem URI

When using Metals with a codebase (discovered when working on the scala-cli codebase), you can get a `publishDiagnostics` message that contains an invalid filesystem URI since it contains `?` making it think it's a query. For example you'll see this in the logs:

```
Mar 27, 2023 12:58:28 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleNotification
WARNING: Notification threw an exception: {
  "jsonrpc": "2.0",
  "method": "build/publishDiagnostics",
  "params": {
    "textDocument": {
      "uri": "file:///Users/ckipp/Documents/scala-workspace/scala-cli/cli/test?id\u003dcli.test"
    },
    "buildTarget": {
      "uri": "file:///Users/ckipp/Documents/scala-workspace/scala-cli/cli/test?id\u003dcli.test"
    },
    "diagnostics": [
      {
        "range": {
          "start": {
            "line": 0,
            "character": 0
          },
          "end": {
            "line": 0,
            "character": 0
          }
        },
        "severity": 2,
        "code": "-1",
        "source": "mill",
        "message": "Flag -Xsemanticdb set repeatedly"
      },
      {
        "range": {
          "start": {
            "line": 0,
            "character": 0
          },
          "end": {
            "line": 0,
            "character": 0
          }
        },
        "severity": 2,
        "code": "-1",
        "source": "mill",
        "message": "Flag -Xsemanticdb set repeatedly"
      }
    ],
    "reset": true
  }
}
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.notify(GenericEndpoint.java:152)
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleNotification(RemoteEndpoint.java:220)
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:187)
        at org.eclipse.lsp4j.jsonrpc.TracingMessageConsumer.consume(TracingMessageConsumer.java:114)
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
        at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
        ... 12 more
Caused by: java.lang.IllegalArgumentException: URI has a query component
        at java.base/sun.nio.fs.UnixUriUtils.fromUri(UnixUriUtils.java:58)
        at java.base/sun.nio.fs.UnixFileSystemProvider.getPath(UnixFileSystemProvider.java:102)
        at java.base/java.nio.file.Path.of(Path.java:203)
        at java.base/java.nio.file.Paths.get(Paths.java:98)
        at scala.meta.internal.mtags.MtagsEnrichments$XtensionURIMtags.toAbsolutePath(MtagsEnrichments.scala:120)
        at scala.meta.internal.mtags.MtagsEnrichments$XtensionStringMtags.toAbsolutePath(MtagsEnrichments.scala:173)
        at scala.meta.internal.metals.MetalsEnrichments$XtensionString.toAbsolutePath(MetalsEnrichments.scala:595)
        at scala.meta.internal.metals.MetalsEnrichments$XtensionString.toAbsolutePath(MetalsEnrichments.scala:592)
        at scala.meta.internal.metals.Diagnostics.onBuildPublishDiagnostics(Diagnostics.scala:142)
        at scala.meta.internal.metals.clients.language.ForwardingMetalsBuildClient.onBuildPublishDiagnostics(ForwardingMetalsBuildClient.scala:109)
        ... 17 more
```

**To Reproduce**
- Clone scala-cli
- Import the build (using mill-bsp)
- See error in logs

**Expected behaviour**
I wouldn't expect the target URI to be used in the `textDocument.uri` here as this isn't a valid filesystem URI.

--- a/bsp/worker/src/mill/bsp/worker/MillBuildServer.scala
+++ b/bsp/worker/src/mill/bsp/worker/MillBuildServer.scala
@@ -58,7 +58,7 @@ import com.google.gson.JsonObject
 import mill.T
 import mill.api.{DummyTestReporter, PathRef, Result, Strict, internal}
 import mill.define.Segment.Label
-import mill.define.{BaseModule, Discover, ExternalModule, Module, Segments, Task}
+import mill.define.{Discover, ExternalModule, Module, Segments, Task}
 import mill.eval.Evaluator
 import mill.main.{BspServerResult, EvaluatorScopt, MainModule}
 import mill.scalalib.{JavaModule, SemanticDbJavaModule, TestModule}
@@ -89,14 +89,14 @@ class MillBuildServer(
 
   lazy val millDiscover: Discover[MillBuildServer.this.type] = Discover[this.type]
 
-  var cancellator: Boolean => Unit = shutdownBefore => ()
-  var onSessionEnd: Option[BspServerResult => Unit] = None
-  var client: BuildClient = _
-  var initialized = false
-  var clientInitialized = false
-  var shutdownRequested = false
-  var clientWantsSemanticDb = false
-  var clientIsIntelliJ = false
+  private[worker] var cancellator: Boolean => Unit = shutdownBefore => ()
+  private[worker] var onSessionEnd: Option[BspServerResult => Unit] = None
+  protected var client: BuildClient = _
+  private var initialized = false
+  private var clientInitialized = false
+  private var shutdownRequested = false
+  protected var clientWantsSemanticDb = false
+  protected var clientIsIntelliJ = false
 
   class State(val evaluator: Evaluator) {
     private[this] object internal {
@@ -107,14 +107,12 @@ class MillBuildServer(
               ModuleUtils.transitiveModules(evaluator.rootModule) ++ Seq(`mill-build`)
             val map = modules.collect {
               case m: MillBuildModule =>
-                val uri = sanitizeUri(m.millSourcePath) +
-                  m.bspBuildTarget.displayName.map(n => s"?id=${n}").getOrElse("")
+                val uri = sanitizeUri(m.millSourcePath)
                 val id = new BuildTargetIdentifier(uri)
                 log.debug(s"mill-build segments: ${`mill-build`.millModuleSegments.render}")
                 (id, m)
               case m: BspModule =>
-                val uri = sanitizeUri(`mill-build`.millSourcePath / m.millModuleSegments.parts) +
-                  m.bspBuildTarget.displayName.map(n => s"?id=${n}").getOrElse("")
+                val uri = sanitizeUri(`mill-build`.millSourcePath / m.millModuleSegments.parts)
                 val id = new BuildTargetIdentifier(uri)
                 (id, m)
             }.toMap
@@ -124,10 +122,8 @@ class MillBuildServer(
           case _ => // already init
         }
       }
-      private[MillBuildServer] var idToModule: Option[Map[BuildTargetIdentifier, BspModule]] =
-        None
+      private[MillBuildServer] var idToModule: Option[Map[BuildTargetIdentifier, BspModule]] = None
       private[MillBuildServer] var modulesToId: Option[Map[BspModule, BuildTargetIdentifier]] = None
-
     }
 
     lazy val `mill-build`: MillBuildModule = {
@@ -218,10 +214,9 @@ class MillBuildServer(
       capabilities.setRunProvider(new RunProvider(supportedLangs))
       capabilities.setTestProvider(new TestProvider(supportedLangs))
 
-      request.getDisplayName match {
-        case "IntelliJ-BSP" => clientIsIntelliJ = true
-        case _ => clientIsIntelliJ = false
-      }
+      // IJ is currently not able to handle files as source paths, only dirs
+      // TODO: Rumor has it, that newer version may handle it, so we need to better detect that
+      clientIsIntelliJ = request.getDisplayName == "IntelliJ-BSP"
 
       def readVersion(json: JsonObject, name: String): Option[String] =
         if (json.has(name)) {
@@ -740,7 +735,9 @@ class MillBuildServer(
 
     if (checkInitialized && !initialized) {
       future.completeExceptionally(
-        new Exception(s"Can not respond to ${prefix} request before receiving the `initialize` request.")
+        new Exception(
+          s"Can not respond to ${prefix} request before receiving the `initialize` request."
+        )
       )
     } else {
       statePromise.future.onComplete {
@@ -780,7 +777,9 @@ class MillBuildServer(
 
     if (checkInitialized && !initialized) {
       future.completeExceptionally(
-        new Exception(s"Can not respond to ${prefix} request before receiving the `initialize` request.")
+        new Exception(
+          s"Can not respond to ${prefix} request before receiving the `initialize` request."
+        )
       )
     } else {
       try {

